buildkite-agent:  secrecy improvements:  non-store, non-Nix-source provisioning of secrets

deepfire opened the pull request:
###### Motivation for this change

1. Remove the requirement to store secrets in Nix source:
   - SSH keys
   - Buildkite token
2. Make sure that the secrets don't hit the Nix store.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

deepfire committed:
buildkite-agent:  secrecy improvements:  non-store, non-Nix provisioning of secrets

deepfire commented:
cc @domenkozar

domenkozar commented:
This is one of efforts for https://github.com/NixOS/nixpkgs/issues/24288

domenkozar commented on nixos/modules/services/continuous-integration/buildkite-agent.nix:
```diff
@@ -21,9 +16,11 @@ in
 
       token = mkOption {
-        type = types.str;
+        type = types.either types.str types.path;
```
I think we should break backwards compatibility here, since the old method was insecure.

We should just allow `types.path`.

deepfire replied:
Done.

domenkozar commented on nixos/modules/services/continuous-integration/buildkite-agent.nix:
```diff
@@ -3,16 +3,11 @@
 
 let
+  ## isPath :: String -> Bool
+  isPath = x: !(isAttrs x || isList x || isFunction x || isString x || isInt x || isBool x || isNull x)
```
Once we use just path, this is redundant (already checked by types.path)

domenkozar replied:
@deepfire needs to be removed

deepfire replied:
Done.

domenkozar commented on nixos/modules/services/continuous-integration/buildkite-agent.nix:
```diff
@@ -65,27 +66,43 @@ in
-
-            echo "${cfg.openssh.publicKey}" > /var/lib/buildkite-agent/.ssh/id_rsa.pub
-            ${pkgs.coreutils}/bin/chmod 600 /var/lib/buildkite-agent/.ssh/id_rsa.pub
+            cat > "/var/lib/buildkite-agent/buildkite-agent.cfg" <<EOF
```
Missing a comment why this is done this way, to prevent issues in the future

deepfire replied:
Done.

domenkozar reviewed (changes requested):
cc @zimbatm

domenkozar commented on nixos/modules/services/continuous-integration/buildkite-agent.nix:
```diff
@@ -72,27 +72,20 @@ in
 
             cat > "/var/lib/buildkite-agent/buildkite-agent.cfg" <<EOF
-            token="${catOrLiteral cfg.token}"
+            token="$(cat ${toString cfg.tokenPath})"
```
One big caveat here is that tokenPath needs to be on the machine running this build, so not something NixOS itself managed (but NixOps does).

I think we should clarify this in option descriptions.

domenkozar commented on nixos/modules/services/continuous-integration/buildkite-agent.nix:
```diff
@@ -15,12 +15,12 @@ in
 
-      token = mkOption {
-        type = types.either types.str types.path;
+      tokenPath = mkOption {
```
The rename can only be done on master, for 17.09 we'll have to keep backwards compatibility :(

For master use `mkRenamedOptionModule` for renaming

deepfire committed:
buildkite-agent:  demotivate potential secrecy regressions through documentation

# Conflicts:
#	nixos/modules/services/continuous-integration/buildkite-agent.nix

deepfire commented:
@domenkozar, dropped the option renaming from this PR -- full version also submitted to master in https://github.com/NixOS/nixpkgs/pull/32009

zimbatm reviewed (commented):
sounds good overall but this should also be ported to master or it's going to disappear in the next release

deepfire commented:
@zimbatm, done!

zimbatm reviewed (approved):
LGTM. @domenkozar and last words? @deepfire please ping me again in a few days if this PR isn't merged

domenkozar commented:
cc @fpletz & @globin
